src/wp-includes/class-wp-theme.php

```diff
@@ -821,6 +821,16 @@ public function cache_delete() {
 		$this->block_template_folders = null;
 		$this->headers                = array();
 		$this->__construct( $this->stylesheet, $this->theme_root );
+		$this->delete_pattern_cache();
+	}
+
+	/**
+	 * Clear block pattern cache.
+	 *
+	 * @since 6.4.0
+	 */
+	public function delete_pattern_cache() {
+		delete_transient( 'wp_theme_patterns_' . $this->stylesheet );
 	}
```

Comment:
Given you added this, I think it would make sense to add the following as well:
* `get_pattern_cache()`: Returns `false` or the array of patterns.
* `set_pattern_cache( array $patterns )`: Sets the patterns in cache.

The two methods could automatically take care of also storing the theme's current version (`set_pattern_cache`) and only returning the cached patterns if the version is the same (`get_pattern_cache`). That would then simplify the logic used in `_wp_get_block_patterns()`.
```suggestion
	/**
	 * Gets patterns from block pattern cache.
	 *
	 * @since 6.4.0
	 *
	 * @return array|false Array with block patterns data, or false if no valid value found in cache.
	 */
	public function get_pattern_cache() {
		$pattern_data = get_transient( 'wp_theme_patterns_' . $this->stylesheet );
		if ( is_array( $pattern_data ) && $pattern_data['version'] === $this->get( 'Version' ) ) {
			return $pattern_data['patterns'];
		}
		return false;
	}

	/**
	 * Sets the block pattern cache.
	 *
	 * @since 6.4.0
	 *
	 * @param array $patterns Block patterns data to store in cache.
	 */
	public function set_pattern_cache( array $patterns ) {
		$pattern_data = array(
			'version'  => $this->get( 'Version' ),
			'patterns' => $patterns,
		);
		set_transient( 'wp_theme_patterns_' . $this->stylesheet, $pattern_data );
	}

	/**
	 * Clears block pattern cache.
	 *
	 * @since 6.4.0
	 */
	public function delete_pattern_cache() {
		delete_transient( 'wp_theme_patterns_' . $this->stylesheet );
	}
```